<a href="https://colab.research.google.com/github/altn714/ESAA/blob/main/%EB%AF%B8%EB%8B%88%ED%94%8C%EC%A0%9D_%ED%8C%8C%EC%83%9D%EB%B3%80%EC%88%98.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. 데이터셋 불러오기

In [138]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [139]:
import numpy as np
import pandas as pd 

In [140]:
train = pd.read_csv("/content/train 지역변수 추가.csv")
test = pd.read_csv("/content/test 지역변수 추가.csv")
sd = pd.read_csv("/content/submission_sample.csv")

In [141]:
train.head(3)

inst_id    OC       sido  sgg  openDate  bedCount          instkind  \
0        1  open  choongnam   73  20071228     175.0  nursing_hospital   
1        3  open  gyeongnam   32  19970401     410.0  general_hospital   
2        4  open   gyeonggi   89  20161228     468.0  nursing_hospital   

       revenue1   salescost1          sga1  ...  liquidLiabilities2  \
0  4.217530e+09          0.0  3.961135e+09  ...         222876855.0   
1           NaN          NaN           NaN  ...                 NaN   
2  1.004522e+09  515483669.0  4.472197e+08  ...                 0.0   

   shortLoan2  NCLiabilities2    longLoan2     netAsset2      surplus2  \
0         0.0     536116887.0  390000000.0  2.619290e+09  1.271224e+09   
1         NaN             NaN          NaN           NaN           NaN   
2         0.0             0.0          0.0  0.000000e+00  0.000000e+00   

   employee1  employee2  ownerChange  hospital_count  
0       62.0       64.0         same            14.0  
1      801.0      813.0         same             5.0  
2      234.0        1.0         same            17.0  

[3 rows x 59 columns]

In [142]:
train.describe()

inst_id         sgg      openDate    bedCount      revenue1  \
count  301.000000  301.000000  3.010000e+02  296.000000  2.930000e+02   
mean   219.056478   81.039867  2.005013e+07  145.709459  1.288175e+10   
std    121.234869   50.969714  8.893815e+04  118.923890  2.043543e+10   
min      1.000000    1.000000  1.978012e+07    0.000000  0.000000e+00   
25%    112.000000   37.000000  2.001102e+07   52.750000  3.252112e+09   
50%    230.000000   75.000000  2.007113e+07  136.500000  5.524218e+09   
75%    321.000000  123.000000  2.011102e+07  193.000000  1.274839e+10   
max    428.000000  178.000000  2.017061e+07  656.000000  1.510000e+11   

         salescost1          sga1       salary1          noi1          noe1  \
count  2.930000e+02  2.930000e+02  2.930000e+02  2.930000e+02  2.930000e+02   
mean   2.014903e+09  1.033244e+10  5.654115e+09  2.696151e+08  5.115870e+08   
std    7.460271e+09  1.493886e+10  8.083343e+09  8.023701e+08  1.060379e+09   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    0.000000e+00  2.758201e+09  1.626053e+09  8.217133e+06  8.013395e+07   
50%    2.104105e+08  4.684074e+09  2.659892e+09  4.363764e+07  1.831965e+08   
75%    9.105278e+08  1.077876e+10  6.363400e+09  2.050331e+08  4.203330e+08   
max    9.850332e+10  1.030000e+11  6.403559e+10  9.144171e+09  8.686380e+09   

       ...         debt2  liquidLiabilities2    shortLoan2  NCLiabilities2  \
count  ...  2.930000e+02        2.930000e+02  2.930000e+02    2.930000e+02   
mean   ...  8.146026e+09        3.860584e+09  1.510050e+09    4.471247e+09   
std    ...  1.255800e+10        6.797242e+09  2.953412e+09    7.659580e+09   
min    ...  0.000000e+00        0.000000e+00  0.000000e+00    0.000000e+00   
25%    ...  1.283220e+09        2.855741e+08  0.000000e+00    2.557878e+07   
50%    ...  3.784553e+09        1.454050e+09  8.542549e+07    1.969747e+09   
75%    ...  8.465053e+09        4.364714e+09  1.567967e+09    4.905441e+09   
max    ...  8.508858e+10        6.846878e+10  1.759375e+10    5.150388e+10   

          longLoan2     netAsset2      surplus2    employee1    employee2  \
count  2.930000e+02  2.930000e+02  2.930000e+02   291.000000   288.000000   
mean   2.709979e+09  5.273919e+09  9.786279e+08   142.546392   134.326389   
std    4.564001e+09  1.081259e+10  4.688798e+09   160.191073   151.061786   
min    0.000000e+00 -2.014780e+10 -2.781507e+09     0.000000     0.000000   
25%    0.000000e+00  1.017573e+09  0.000000e+00    53.500000    53.750000   
50%    1.100000e+09  2.894970e+09  0.000000e+00    80.000000    79.000000   
75%    3.360000e+09  5.370285e+09  1.014447e+08   181.500000   170.000000   
max    3.256147e+10  1.250000e+11  6.852730e+10  1200.000000  1200.000000   

       hospital_count  
count      299.000000  
mean        10.230769  
std          6.221061  
min          1.000000  
25%          5.000000  
50%         10.000000  
75%         15.500000  
max         20.000000  

[8 rows x 55 columns]

In [143]:
test.head(3)

inst_id  OC     sido  sgg    openDate  bedCount          instkind  \
0        2 NaN  incheon  139  19981125.0     300.0  general_hospital   
1        5 NaN     jeju  149  20160309.0      44.0          hospital   
2        6 NaN  jeonnam  103  19890427.0     276.0  general_hospital   

       revenue1    salescost1          sga1  ...  liquidLiabilities2  \
0  6.682486e+10  0.000000e+00  6.565709e+10  ...        5.068443e+10   
1  3.495758e+10  0.000000e+00  3.259270e+10  ...        4.209828e+10   
2  2.326031e+10  2.542571e+09  2.308749e+10  ...        2.777589e+10   

     shortLoan2  NCLiabilities2     longLoan2     netAsset2      surplus2  \
0  3.714334e+10    4.720000e+09  4.690000e+09  1.608540e+10  8.944587e+09   
1  2.420000e+10    2.521009e+10  1.830000e+10  3.789135e+09  0.000000e+00   
2  2.182278e+10    0.000000e+00  0.000000e+00  0.000000e+00  1.638540e+10   

   employee1  employee2  ownerChange  hospital_count  
0      693.0      693.0         same             1.0  
1      379.0      371.0         same             1.0  
2        NaN        NaN          NaN             3.0  

[3 rows x 59 columns]

### *변수설명
inst_id - 각 파일에서의 병원 고유 번호   
OC – 영업/폐업 분류, 2018년 폐업은 2017년 폐업으로 간주함    
sido – 병원의 광역 지역 정보    
sgg – 병원의 시군구 자료     
openDate – 병원 설립일    
bedCount - 병원이 갖추고 있는 병상의 수       
instkind – 병원, 의원, 요양병원, 한의원, 종합병원 등 병원의 종류    
·        종합병원 : 입원환자 100명 이상 수용 가능    
·        병원 : 입원 환자 30명 이상 100명 미만 수용 가능     
·        의원 : 입원 환자 30명 이하 수용 가능      
·        한방 병원(한의원) : 침술과 한약으로 치료하는 의료 기관.    
revenue1 – 매출액 / salescost1 – 매출원가      
sga1 - 판매비와 관리비 / salary1 – 급여   
noi1 – 영업외수익 / noe1 – 영업외비용      
Interest1 – 이자비용 / ctax1 – 법인세비용     
Profit1 – 당기순이익 / liquidAsset1 – 유동자산      
quickAsset1 – 당좌자산 / receivableS1 - 미수금(단기)       
inventoryAsset1 – 재고자산 / nonCAsset1 – 비유동자산     
tanAsset1 – 유형자산 / OnonCAsset1 - 기타 비유동자산       
receivableL1 – 장기미수금 / debt1 – 부채총계      
liquidLiabilities1 – 유동부채 / shortLoan1 – 단기차입금    
NCLiabilities1 – 비유동부채 / longLoan1 – 장기차입금    
netAsset1 – 순자산총계 / surplus1 – 이익잉여금    
employee1 – 고용한 총 직원의 수     
ownerChange – 대표자의 변동 

### 1118회의내용
[전처리]
1. inst_id, sgg 빼기
2. instkind는 결측치 삭제. 1개밖에 안됨
3. 숫자형,문자형 결측치는 최빈값으로? 자료더살펴보기

[파생변수만들기]
1. 재무비율
2. 병상수 300이상이면 대형병원, 300미만이면 중소병원
3. 지역x병원종류 수 -> 경쟁적 환경


# 2. EDA 및 전처리

In [144]:
train.isnull().sum()

inst_id                0
OC                     0
sido                   0
sgg                    0
openDate               0
bedCount               5
instkind               1
revenue1               8
salescost1             8
sga1                   8
salary1                8
noi1                   8
noe1                   8
interest1              8
ctax1                  8
profit1                8
liquidAsset1           8
quickAsset1            8
receivableS1           8
inventoryAsset1        8
nonCAsset1             8
tanAsset1              8
OnonCAsset1            8
receivableL1           8
debt1                  8
liquidLiabilities1     8
shortLoan1             8
NCLiabilities1         8
longLoan1              8
netAsset1              8
surplus1               8
revenue2               8
salescost2             8
sga2                   8
salary2                8
noi2                   8
noe2                   8
interest2              8
ctax2                  8
profit2                8


In [145]:
test.isnull().sum()

inst_id                 0
OC                    127
sido                    0
sgg                     0
openDate                1
bedCount                8
instkind                2
revenue1                2
salescost1              2
sga1                    2
salary1                 2
noi1                    2
noe1                    2
interest1               2
ctax1                   2
profit1                 2
liquidAsset1            2
quickAsset1             2
receivableS1            2
inventoryAsset1         2
nonCAsset1              2
tanAsset1               2
OnonCAsset1             2
receivableL1            2
debt1                   2
liquidLiabilities1      2
shortLoan1              2
NCLiabilities1          2
longLoan1               2
netAsset1               2
surplus1                2
revenue2                2
salescost2              2
sga2                    2
salary2                 2
noi2                    2
noe2                    2
interest2               2
ctax2       

### (1) 레이블 인코딩

In [146]:
from sklearn.preprocessing import LabelEncoder

# OC : 'close': 0, 'open': 1 처리해주기
print(train['OC'].value_counts()) #인코딩전 확인
encoder1 = LabelEncoder()
train['OC']=encoder1.fit_transform(train['OC'])
print(train['OC'].value_counts()) #인코딩후 확인

# 테스트데이터셋은 OC가 모두 NaN이므로 레이블인코딩 처리x

open      286
 close     15
Name: OC, dtype: int64
1    286
0     15
Name: OC, dtype: int64


### (2) 피처 결측치 처리

#### (2)-1. ints_id, sgg

In [147]:
# inst_id : 인덱스 중복이므로 열 버리기
# sgg : sido와 정보가 겹치므로 열 버리기
train.drop(['inst_id','sgg'],axis=1,inplace=True)
test.drop(['inst_id','sgg'],axis=1,inplace=True)

#### (2)-2. 문자열 피처

In [148]:
## 문자열 피처중 instkind, ownerChange 에서 결측치 존재

# instkind(문자열) : 병원종류 결측치 있는 행 다지우기
train.dropna(subset=['instkind'],axis=0,inplace=True)
test.dropna(subset=['instkind'],axis=0,inplace=True)

In [149]:
#ownerChange 'same'과 'change'의 전체비율 구하기
round(train['ownerChange'].value_counts(normalize=True))

same      1.0
change    0.0
Name: ownerChange, dtype: float64

In [150]:
round(train['ownerChange'].value_counts(normalize=True),3)

# ownerChange(문자열) : 전체 비율로 대체 *문자형 숫자형 대체
train.loc[train.ownerChange=='same','ownerChange']=0.87
train.loc[train.ownerChange=='change','ownerChange']=0.13

test.loc[test.ownerChange=='same','ownerChange']=0.87
test.loc[test.ownerChange=='change','ownerChange']=0.13

In [151]:
train.isnull().sum()

OC                     0
sido                   0
openDate               0
bedCount               4
instkind               0
revenue1               8
salescost1             8
sga1                   8
salary1                8
noi1                   8
noe1                   8
interest1              8
ctax1                  8
profit1                8
liquidAsset1           8
quickAsset1            8
receivableS1           8
inventoryAsset1        8
nonCAsset1             8
tanAsset1              8
OnonCAsset1            8
receivableL1           8
debt1                  8
liquidLiabilities1     8
shortLoan1             8
NCLiabilities1         8
longLoan1              8
netAsset1              8
surplus1               8
revenue2               8
salescost2             8
sga2                   8
salary2                8
noi2                   8
noe2                   8
interest2              8
ctax2                  8
profit2                8
liquidAsset2           8
quickAsset2            8


#### (2)-3. 숫자형 피처중 금융변수

In [152]:
## 숫자형 피처중 금융변수에서 결측치 존재

# open&같은행번호에서 공통적으로 8개의 결측치가 나타난다. 
print(train[train['revenue1'].isnull()]['OC'])
print(train[train['shortLoan2'].isnull()]['OC'])

# open&같은행번호에서 공통적으로 2개의 결측치가 나타난다.
print(test[test['revenue1'].isnull()])
print(test[test['shortLoan2'].isnull()])

# 따라서 결측치가 존재하는 전체행 지우기
train.dropna(subset=['revenue1'],axis=0,inplace=True)
test.dropna(subset=['revenue1'],axis=0,inplace=True)

1      1
14     1
32     1
55     1
60     1
248    1
257    1
258    1
Name: OC, dtype: int64
1      1
14     1
32     1
55     1
60     1
248    1
257    1
258    1
Name: OC, dtype: int64
    OC     sido    openDate  bedCount          instkind  revenue1  salescost1  \
15 NaN    busan  19950330.0     226.0  general_hospital       NaN         NaN   
56 NaN  jeonbuk  20121218.0     539.0  nursing_hospital       NaN         NaN   

    sga1  salary1  noi1  ...  liquidLiabilities2  shortLoan2  NCLiabilities2  \
15   NaN      NaN   NaN  ...                 NaN         NaN             NaN   
56   NaN      NaN   NaN  ...                 NaN         NaN             NaN   

    longLoan2  netAsset2  surplus2  employee1  employee2  ownerChange  \
15        NaN        NaN       NaN      566.0      566.0         0.87   
56        NaN        NaN       NaN      250.0      200.0         0.87   

    hospital_count  
15             2.0  
56             6.0  

[2 rows x 57 columns]
    OC     sido    

In [153]:
# 전체데이터 중간확인
print(train.head(3))
print(train.isnull().sum())

   OC       sido  openDate  bedCount          instkind      revenue1  \
0   1  choongnam  20071228     175.0  nursing_hospital  4.217530e+09   
2   1   gyeonggi  20161228     468.0  nursing_hospital  1.004522e+09   
3   1    incheon  20000814     353.0  general_hospital  7.250734e+10   

    salescost1          sga1       salary1         noi1  ...  \
0          0.0  3.961135e+09  2.033835e+09   15652441.0  ...   
2  515483669.0  4.472197e+08  2.964023e+08      76156.0  ...   
3          0.0  7.067740e+10  3.178605e+10  506223059.0  ...   

   liquidLiabilities2    shortLoan2  NCLiabilities2     longLoan2  \
0        2.228769e+08  0.000000e+00    5.361169e+08  3.900000e+08   
2        0.000000e+00  0.000000e+00    0.000000e+00  0.000000e+00   
3        1.701860e+10  9.219427e+09    2.073641e+10  1.510000e+10   

      netAsset2      surplus2  employee1  employee2  ownerChange  \
0  2.619290e+09  1.271224e+09       62.0       64.0         0.87   
2  0.000000e+00  0.000000e+00      234.0 

In [154]:
# 전체데이터 중간확인
print(test.isnull().sum())

OC                    123
sido                    0
openDate                0
bedCount                7
instkind                0
revenue1                0
salescost1              0
sga1                    0
salary1                 0
noi1                    0
noe1                    0
interest1               0
ctax1                   0
profit1                 0
liquidAsset1            0
quickAsset1             0
receivableS1            0
inventoryAsset1         0
nonCAsset1              0
tanAsset1               0
OnonCAsset1             0
receivableL1            0
debt1                   0
liquidLiabilities1      0
shortLoan1              0
NCLiabilities1          0
longLoan1               0
netAsset1               0
surplus1                0
revenue2                0
salescost2              0
sga2                    0
salary2                 0
noi2                    0
noe2                    0
interest2               0
ctax2                   0
profit2                 0
liquidAsset2

#### (2)-4. bedcount 
instkind별 평균값을 bedcount의 결측값으로 처리

In [155]:
train.groupby('instkind')['bedCount'].median()
train['bedCount'] = train['bedCount'].fillna(train.groupby('instkind')['bedCount'].transform('median')) 

train['bedCount'].isnull().sum()

0

In [156]:
test.groupby('instkind')['bedCount'].median()
test['bedCount'] = test['bedCount'].fillna(test.groupby('instkind')['bedCount'].transform('median')) 
test['bedCount'].isnull().sum()
print(test['bedCount'])

0      300.0
1       44.0
2      276.0
3      363.0
4      213.0
       ...  
121    167.0
122      0.0
123      NaN
124    150.0
126     96.0
Name: bedCount, Length: 123, dtype: float64


**test 파일에서 instkind별 bedCount 중앙값 결측치 처리 -> 123행 결측치 처리 안됨. (이유: traditional_hospital의 bedCount 행 하나밖에 없음.)**

In [157]:
#결측치 행 추출
test[test['bedCount'].isnull()]

OC      sido    openDate  bedCount              instkind      revenue1  \
123 NaN  gyeonggi  20010530.0       NaN  traditional_hospital  1.833906e+10   

     salescost1          sga1       salary1        noi1  ...  \
123         0.0  1.760117e+10  6.824241e+09  31454472.0  ...   

     liquidLiabilities2  shortLoan2  NCLiabilities2     longLoan2  \
123        1.376742e+09         0.0    1.799886e+09  1.740000e+09   

        netAsset2      surplus2  employee1  employee2  ownerChange  \
123  8.574502e+09  7.136681e+09      132.0      137.0         0.87   

     hospital_count  
123             1.0  

[1 rows x 57 columns]

#### (2)-5. employee1,2 결측치 중앙값 대체

In [158]:
#employee1 결측치 중앙값 대체 
train['employee1'].median()
train.loc[train['employee1'].isna(),'employee1'] = train['employee1'].median()
train['employee1'].isna().sum()

test['employee1'].median()
test.loc[test['employee1'].isna(),'employee1'] = test['employee1'].median()
test['employee1'].isna().sum()

0

In [159]:
#employee2 결측치 중앙값 대체 
train['employee2'].median()
train.loc[train['employee2'].isna(),'employee2'] = train['employee2'].median()
train['employee2'].isna().sum()

test['employee2'].median()
test.loc[test['employee2'].isna(),'employee2'] = test['employee2'].median()
test['employee2'].isna().sum()

0

* 매출액 증가율 = 당기매출액 / 전기매출액 *100 -100
* 총자산 증가율 = 당기총자산 / 전기총자산 *100 -100
* 자기자본순이익률(ROE) = 당기순이익/자기자본*100


In [160]:
# 매출액 증가율 = 당기매출액 / 전기매출액 *100 -100
#                  revenue1 / revenue2
# 총자산 증가율 = 당기총자산 / 전기총자산 *100 -100
#            (netAsset1+debt1)/(netAsset2+debt2)
# 순자산총계 = 자산총계-부채총계
# 자산 = 부채+자본
# 자산총계 = 순자산총계 + 부채총계

# 자기자본순이익률(ROE) = 당기순이익/총자본*100
#                          profit1 / (netAsset1) 
# 자기자본 = 자본금+이익잉여금

In [161]:
train['매출액증가율']=(train['revenue1']/train['revenue2'])*100-100
train['총자산증가율']=(train['netAsset1']+train['debt1'])/(train['netAsset2']+train['debt2'])*100 -100
train['ROE']=(train['profit1']/train['netAsset1'])*100

In [162]:
print(train['매출액증가율'])
print(train['총자산증가율'])
print(train['ROE'])

0       -1.868801
2             inf
3        8.449217
4        1.998089
5       -2.195782
          ...    
296      7.722377
297           inf
298   -100.000000
299    -48.329060
300    -10.543000
Name: 매출액증가율, Length: 292, dtype: float64
0       4.410580
2            inf
3      10.869051
4      -2.826186
5            NaN
         ...    
296    -0.770201
297          NaN
298    -1.997912
299    -9.199578
300     0.220830
Name: 총자산증가율, Length: 292, dtype: float64
0       7.916078
2      13.931289
3       6.515289
4      93.555978
5           -inf
         ...    
296     9.949747
297         -inf
298    -2.251142
299         -inf
300    -0.828848
Name: ROE, Length: 292, dtype: float64


In [163]:
train['netAsset1']

0      2.844460e+09
2      3.005088e+08
3      1.385710e+10
4      8.631164e+07
5      0.000000e+00
           ...     
296    5.973889e+09
297    0.000000e+00
298    3.187408e+09
299    0.000000e+00
300    5.490668e+09
Name: netAsset1, Length: 292, dtype: float64

<비금융 파생변수> 
A. 중소병원 병상수>300 : 대형병원 1 병상수<=300 : 중소병원 0 중소병원일수록 폐업률이 높아진다.

In [164]:
train.loc[train.bedCount>300,'hospital_size']=1
train.loc[train.bedCount<=300,'hospital_size']=0

금융변수

In [165]:
## *성장성
# 매출액 증가율 = 당기매출액/ 전기매출액*100 -100
train['매출액증가율']=(train['revenue1']/train['revenue2'])*100-100
# 총자산 증가율 = 당기총자산/ 전기총자산*100 -100
train['총자산증가율']=(train['netAsset1']+train['debt1'])/(train['netAsset2']+train['debt2'])*100 -100

## 수익성
# 자기자본순이익률(ROE)=당기순이익/자기자본*100
train['ROE']=(train['profit1']/train['netAsset1'])*100
# 매출순이익률 = 당기순이익/매출액*100
train['매출액순이익율'] = train.profit1/train.revenue1*100
# 배당성향 = 배당금/당기순이익
train['배당성향'] = (train.profit1-train.surplus1)/train.profit1
# 이자보상비율 = 영업이익/이자비용*100
# (영업손익은 매출액 등의 영업수익으로부터 매출원가ㆍ판매비와관리비 등의 영업비용을 차감한 개념, 영업 외 이익과 비용도 차감)
train['이자보상비율'] = (train.revenue1-train.salescost1-train.sga1-train.noi1-train.noe1)/train.interest1*100

## 안전성
# 유동비율 = 유동자산/유동부채*100 (바람직: 200%이상)
train['유동비율']=(train['liquidAsset1']/train['liquidLiabilities1'])*100
# 당좌비율 = 당좌자산/유동부채*100 (바람직: 100%이상)
train['당좌비율']=(train['quickAsset1']/train['liquidLiabilities1'])*100
# 부채비율 = 부채/자본*100
train['부채비율']=(train['debt1']/(train['debt1']+train['netAsset1']))*100
# 자기자본비율 = 자기자본/총자본*100
train['자기자본비율']=train.netAsset1/(train.debt1+train.netAsset1)*100

## 활동성
# 재고자산회전율 = 매출원가/평균재고자산
train['재고자산회전율']=train.salescost1/((train.inventoryAsset1+train.inventoryAsset2)/2)

원핫인코딩 : sido, instkind 피처 시키기. train, test 파일 다 시키기.

In [166]:
train=pd.get_dummies(data = train, columns = ['sido'], prefix = 'sido')
train= pd.get_dummies(data = train, columns = ['instkind'], prefix = 'instkind')


test= pd.get_dummies(data = test, columns = ['sido'], prefix = 'sido')
test=pd.get_dummies(data = test, columns = ['instkind'], prefix = 'instkind')

In [178]:
train.drop(columns = 'sido',axis=1)
train.drop(columns = 'instkind',axis=1)

KeyError: ignored

In [181]:
train.columns

Index(['OC', 'openDate', 'bedCount', 'revenue1', 'salescost1', 'sga1',
       'salary1', 'noi1', 'noe1', 'interest1', 'ctax1', 'profit1',
       'liquidAsset1', 'quickAsset1', 'receivableS1', 'inventoryAsset1',
       'nonCAsset1', 'tanAsset1', 'OnonCAsset1', 'receivableL1', 'debt1',
       'liquidLiabilities1', 'shortLoan1', 'NCLiabilities1', 'longLoan1',
       'netAsset1', 'surplus1', 'revenue2', 'salescost2', 'sga2', 'salary2',
       'noi2', 'noe2', 'interest2', 'ctax2', 'profit2', 'liquidAsset2',
       'quickAsset2', 'receivableS2', 'inventoryAsset2', 'nonCAsset2',
       'tanAsset2', 'OnonCAsset2', 'receivableL2', 'debt2',
       'liquidLiabilities2', 'shortLoan2', 'NCLiabilities2', 'longLoan2',
       'netAsset2', 'surplus2', 'employee1', 'employee2', 'ownerChange',
       'hospital_count', '매출액증가율', '총자산증가율', 'ROE', 'hospital_size', '매출액순이익율',
       '배당성향', '이자보상비율', '유동비율', '당좌비율', '부채비율', '자기자본비율', '재고자산회전율',
       'sido_busan', 'sido_choongbuk', 'sido_choongnam', 'sido_d

In [167]:
train

OC  openDate  bedCount      revenue1    salescost1          sga1  \
0     1  20071228     175.0  4.217530e+09  0.000000e+00  3.961135e+09   
2     1  20161228     468.0  1.004522e+09  5.154837e+08  4.472197e+08   
3     1  20000814     353.0  7.250734e+10  0.000000e+00  7.067740e+10   
4     1  20050901     196.0  4.904354e+10  0.000000e+00  4.765605e+10   
5     1  20020501     243.0  3.358054e+10  9.222997e+09  2.372791e+10   
..   ..       ...       ...           ...           ...           ...   
296   1  20050211     214.0  7.614697e+09  4.153475e+08  5.903119e+09   
297   0  20140618      72.5  6.717144e+09  8.200000e+09  0.000000e+00   
298   0  19830705      72.5  0.000000e+00  0.000000e+00  5.479428e+07   
299   0  20070101     180.0  2.116892e+09  2.681748e+08  3.286245e+09   
300   1  20011017      44.0  1.340971e+09  8.108450e+08  5.043409e+08   

          salary1         noi1          noe1     interest1  ...  sido_sejong  \
0    2.033835e+09   15652441.0  1.523624e+07  1.323624e+07  ...            0   
2    2.964023e+08      76156.0  3.000000e+04  0.000000e+00  ...            0   
3    3.178605e+10  506223059.0  1.259568e+09  1.196881e+09  ...            0   
4    2.446078e+10  112352259.0  1.419089e+09  1.307249e+09  ...            0   
5    1.665533e+10   65414325.0  7.412694e+08  3.364622e+08  ...            0   
..            ...          ...           ...           ...  ...          ...   
296  3.467861e+09   42098974.0  7.439421e+08  1.875761e+08  ...            0   
297  0.000000e+00     186359.0  1.398513e+08  1.296975e+08  ...            0   
298  7.500000e+06  121022160.0  1.379810e+08  1.379810e+08  ...            0   
299  2.231944e+09  123665065.0  5.366919e+08  5.143688e+08  ...            0   
300  1.318978e+08    2264007.0  7.355850e+07  7.355850e+07  ...            0   

     sido_seoul  sido_ulsan  instkind_clinic  instkind_dental_clinic  \
0             0           0                0                       0   
2             0           0                0                       0   
3             0           0                0                       0   
4             0           0                0                       0   
5             0           0                0                       0   
..          ...         ...              ...                     ...   
296           0           0                0                       0   
297           0           0                0                       0   
298           0           0                0                       0   
299           0           0                0                       0   
300           1           0                0                       0   

     instkind_general_hospital  instkind_hospital  instkind_nursing_hospital  \
0                            0                  0                          1   
2                            0                  0                          1   
3                            1                  0                          0   
4                            1                  0                          0   
5                            1                  0                          0   
..                         ...                ...                        ...   
296                          0                  0                          1   
297                          0                  1                          0   
298                          0                  1                          0   
299                          0                  0                          1   
300                          0                  0                          0   

     instkind_traditional_clinic  instkind_traditional_hospital  
0                              0                              0  
2                              0                              0  
3                              0                              0  
4                              0                              0  
5          

In [173]:
X_features = train.drop(columns = 'OC',axis=1)
y_labels = train['OC']

In [174]:
X_train, X_test, y_train, y_test = train_test_split(X_features, y_labels, test_size=0.2, random_state=156)

In [176]:
# LightGBM의 파이썬 패키지인 lightgbm에서 LGBMClassifier 임포트
from lightgbm import LGBMClassifier
from sklearn.model_selection import train_test_split
import warnings
warnings.filterwarnings('ignore')

In [177]:
# 앞서 XGBoost와 동일하게 n_estimators는 400설정
lgbm_wrapper = LGBMClassifier(n_estimators=400)

# LightGBM도 XGBoost와 동일하게 조기중단 수행가능
evals = [(X_test, y_test)]
lgbm_wrapper.fit(X_train, y_train,
                 early_stopping_rounds=100, eval_metric="logloss",
                 eval_set=evals, verbose=True)
preds = lgbm_wrapper.predict(X_test)
pred_proba = lgbm_wrapper.predict_proba(X_test)[:, 1]

ValueError: ignored

In [182]:
type(train['ownerChange'])

pandas.core.series.Series

In [186]:
train['ownerChange'].value_counts()

0.87    241
0.13     41
Name: ownerChange, dtype: int64

In [188]:
train[train['ownerChange'].isnull()]

OC  openDate  bedCount      revenue1    salescost1          sga1  \
48    1  19940425      28.0  8.872427e+09  9.641249e+08  7.638920e+09   
62    1  20170602     128.0  5.236406e+08  2.306773e+07  8.355775e+08   
93    0  20070420     203.0  5.423357e+09  0.000000e+00  5.212041e+09   
206   1  20170120     160.0  3.279139e+09  3.726966e+08  3.507972e+09   
212   0  19990603      66.0  2.233031e+10  8.484657e+08  1.849255e+10   
242   1  20170420       0.0  6.845503e+08  0.000000e+00  5.619356e+08   
263   1  20020123      39.0  3.007769e+09  1.197756e+09  2.258240e+09   
285   1  19881104     119.0  9.618709e+09  0.000000e+00  9.751211e+09   
298   0  19830705      72.5  0.000000e+00  0.000000e+00  5.479428e+07   
299   0  20070101     180.0  2.116892e+09  2.681748e+08  3.286245e+09   

          salary1         noi1          noe1    interest1  ...  sido_sejong  \
48   4.488958e+09   19153521.0  2.160889e+08  204479315.0  ...            0   
62   5.644636e+08    7534205.0  5.263060e+07   52630597.0  ...            0   
93   2.958250e+09   23843435.0  1.288886e+08  128638631.0  ...            0   
206  2.659892e+09    1596568.0  8.306566e+07   83065655.0  ...            0   
212  1.232241e+10  423628901.0  1.304817e+09  417949094.0  ...            0   
242  3.014000e+08    8067090.0  2.748281e+06    1364824.0  ...            0   
263  1.775322e+09  611489427.0  5.279657e+07   52076570.0  ...            0   
285  4.979714e+09  813200615.0  5.543432e+08   27044684.0  ...            0   
298  7.500000e+06  121022160.0  1.379810e+08  137980960.0  ...            0   
299  2.231944e+09  123665065.0  5.366919e+08  514368798.0  ...            0   

     sido_seoul  sido_ulsan  instkind_clinic  instkind_dental_clinic  \
48            1           0                0                       0   
62            0           0                0                       0   
93            0           0                0                       0   
206           0           0                0                       0   
212           0           0                0                       0   
242           0           0                0                       0   
263           0           0                0                       0   
285           0           0                0                       0   
298           0           0                0                       0   
299           0           0                0                       0   

     instkind_general_hospital  instkind_hospital  instkind_nursing_hospital  \
48                           0                  0                          0   
62                           0                  0                          1   
93                           0                  1                          0   
206                          0                  0                          1   
212                          0                  0                          1   
242                          0                  0                          0   
263                          0                  1                          0   
285                          0                  1                          0   
298                          0                  1                          0   
299                          0                  0                          1   

     instkind_traditional_clinic  instkind_traditional_hospital  
48                             0                              1  
62                             0                              0  
93                             0                              0  
206                            0                              0  
212                            0                              0  
242                            1                              0  
263                            0                              0  
285                            0                              0  
298                            0                              0

In [191]:
train['ownerChange'].isna().sum()

10